[iOS] Add screen orientation locking / unlocking support outside layout tests

cdumez opened the pull request:
#### d6f850a7728e6788cfd4947b40ca5ec45976bc7a
<pre>
[iOS] Add screen orientation locking / unlocking support outside layout tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=246435">https://bugs.webkit.org/show_bug.cgi?id=246435</a>

Reviewed by Wenson Hsieh.

Add screen orientation locking / unlocking support outside layout tests, on iOS
devices. Outside tests, the fullscreen API uses the WKFullScreenWindowController
to display a new fullscreen view on top of the current one. This new view has its
own UIViewController that WebKit controls. As a result, when locking the screen
orientation (which is gated on being in fullscreen), we need to do so via the
WKFullScreenWindowController, instead of relying on the client application.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::lockFullscreenOrientation):
(WebKit::WebFullScreenManagerProxy::unlockFullscreenOrientation):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
(WebKit::WebFullScreenManagerProxyClient::lockFullscreenOrientation):
(WebKit::WebFullScreenManagerProxyClient::unlockFullscreenOrientation):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
(WebKit::WebScreenOrientationManagerProxy::lock):
(WebKit::WebScreenOrientationManagerProxy::unlock):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::toUIInterfaceOrientationMask):
(WebKit::PageClientImpl::lockFullscreenOrientation):
(WebKit::PageClientImpl::unlockFullscreenOrientation):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController initWithWebView:]):
(-[WKFullScreenViewController setSupportedOrientation:]):
(-[WKFullScreenViewController supportedInterfaceOrientations]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController setSupportedOrientations:]):

Canonical link: <a href="https://commits.webkit.org/255555@main">https://commits.webkit.org/255555@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e05e3802d107a5be6236864cf6323d110dbc6e01

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92874 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23461 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102599 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2088 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30427 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85274 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98538 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79361 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28344 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83341 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36828 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34627 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18158 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/3845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38498 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1773 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40414 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37338 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -741,6 +742,32 @@ - (BOOL)_isHostedInAnotherProcess;
+    case WebCore::ScreenOrientationType::LandscapeSecondary:
+        return UIInterfaceOrientationMaskLandscapeRight;
+    }
+    return UIInterfaceOrientationMaskPortrait;
```
Nit - `ASSERT_NOT_REACHED()`?

cdumez replied:
Ok.

whsieh commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -180,6 +181,17 @@ - (void)dealloc
+
+- (UIInterfaceOrientationMask)supportedInterfaceOrientations
+{
+    return _supportedOrientations;
```
Should this call into the superclass? From looking at `UIViewController`'s logic, it seems that in the `UIUserInterfaceIdiomPhone` case, we use `UIInterfaceOrientationMaskAllButUpsideDown` by default.

cdumez replied:
We can return UIInterfaceOrientationMaskAllButUpsideDown on iPhone instead of UIInterfaceOrientationMaskAll. That said, based on local testing returning `UIInterfaceOrientationMaskAll` on iPhone still doesn't allow you to make the orientation upside down.

I don't understand the comment about calling into the base class. Do you mean that _supportedOrientations would be a std::optional (or whatever the ObjC equivalent of that) and we'd call into the base class if std::nullopt (meaning orientation is unlocked)?

Right now, _supportedOrientations is set unconditionally so...

whsieh replied:
I meant, defaulting `_supportedOrientations` to the value of `[super supportedInterfaceOrientations]` rather than `UIInterfaceOrientationMaskAll`, since UIKit may be choose to be more restrictive than `UIInterfaceOrientationMaskAll` in some user interface idioms.

That said, perhaps it doesn't make any practical difference if (in your testing) it doesn't affect iPhone behavior. Maybe there's more enforcement further down the stack (SpringBoard, BackBoard, etc.)

cdumez committed:
[iOS] Add screen orientation locking / unlocking support outside layout tests
https://bugs.webkit.org/show_bug.cgi?id=246435

Reviewed by Wenson Hsieh.

Add screen orientation locking / unlocking support outside layout tests, on iOS
devices. Outside tests, the fullscreen API uses the WKFullScreenWindowController
to display a new fullscreen view on top of the current one. This new view has its
own UIViewController that WebKit controls. As a result, when locking the screen
orientation (which is gated on being in fullscreen), we need to do so via the
WKFullScreenWindowController, instead of relying on the client application.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::lockFullscreenOrientation):
(WebKit::WebFullScreenManagerProxy::unlockFullscreenOrientation):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
(WebKit::WebFullScreenManagerProxyClient::lockFullscreenOrientation):
(WebKit::WebFullScreenManagerProxyClient::unlockFullscreenOrientation):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
(WebKit::WebScreenOrientationManagerProxy::lock):
(WebKit::WebScreenOrientationManagerProxy::unlock):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::toUIInterfaceOrientationMask):
(WebKit::PageClientImpl::lockFullscreenOrientation):
(WebKit::PageClientImpl::unlockFullscreenOrientation):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController initWithWebView:]):
(-[WKFullScreenViewController setSupportedOrientation:]):
(-[WKFullScreenViewController supportedInterfaceOrientations]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController setSupportedOrientations:]):

Canonical link: https://commits.webkit.org/255555@main
